table_engine/src/partition/rule.rs

```diff
@@ -18,91 +17,55 @@ pub enum Error {
 
     #[snafu(display("Only support one hash column.\nBacktrace:\n{}", backtrace))]
     TooMuchHashColumn { backtrace: Backtrace },
-
-    #[snafu(display("Failed to eval partition expr, err:{}", source))]
-    EvalExpr {
-        source: crate::partition::expression::Error,
-    },
 }
 
 define_result!(Error);
 
 /// Partition rule locate partition by input records
-// TODO(yingwen): Recreate partition rule once the schema of the table is changed
-#[derive(Debug)]
-pub enum PartitionRule {
-    None,
-    Hash(HashPartitionRule),
-}
-
-impl PartitionRule {
-    pub fn new(partition_info: &PartitionInfo, schema: &Schema) -> Result<Self> {
-        match partition_info.partition_type {
-            PartitionType::None => Ok(PartitionRule::None),
-            PartitionType::Hash => {
-                let rule = HashPartitionRule::new(partition_info, schema)?;
-                Ok(PartitionRule::Hash(rule))
-            }
-        }
-    }
-
-    /// Return the index of partition
-    pub fn locate_partition(&self, row: &Row) -> Result<usize> {
-        match self {
-            // Always return the first partition
-            PartitionRule::None => Ok(0),
-            PartitionRule::Hash(rule) => rule.eval_partition_index(row),
-        }
-    }
+pub trait PartitionRule {
+    fn locate_partitions(&self, exprs: &[Expr]) -> Result<Vec<usize>>;
 }
 
 /// Partition rule based on hash
+///
+/// ```SQL
+/// CREATE TABLE employees (
+///     id INT NOT NULL,
+///     fname VARCHAR(30)
+/// )
+/// PARTITION BY HASH(id)
+/// PARTITIONS 4;
+/// ```
+///
+/// Refer to https://dev.mysql.com/doc/refman/8.0/en/partitioning-hash.html to get more information.
+// TODO: Support hash by expression such as `PARTITION BY HASH(col1 + col2)`.
 #[derive(Debug)]
+#[allow(dead_code)]
 pub struct HashPartitionRule {
+    /// Partition column
+    columns: Vec<String>,
     /// Total number of partitions
     partition_num: u16,
-    /// Expression to evaluate a hash value
-    expression: Expression,
-    /// Offsets of columns for evaluate
-    // TODO(yingwen): The column index may be invalid after schema change (add/del column)
-    column_index: SmallVec<[usize; HASH_COLUMN_NUM]>,
 }
 
-impl HashPartitionRule {
-    pub fn new(partition_info: &PartitionInfo, schema: &Schema) -> Result<Self> {
-        let expr = Expression::new(partition_info);
-
-        let col_name_list = expr.extract_column_name();
-        let mut column_index = SmallVec::with_capacity(col_name_list.size_hint().0);
-        for col_name in col_name_list {
-            for (i, v) in schema.columns().iter().enumerate() {
-                if col_name == v.name {
-                    column_index.push(i);
-                    break;
-                }
-            }
-        }
+impl PartitionRule for HashPartitionRule {
```

Comment:
`PartitionRule` should be place in a new pr, and the methods in it should be considered more.